gene2phenotype_project/gene2phenotype_app/serializers/locus_genotype_disease.py

```diff
@@ -447,6 +452,134 @@ def update(self, instance, validated_data):
+
+        """
+        molecular_mechanism_value = validated_data.get("molecular_mechanism")["name"]
+        molecular_mechanism_support = validated_data.get("molecular_mechanism")["support"]
+        mechanism_synopsis = validated_data.get("mechanism_synopsis", None) # synopsis is optional
+        mechanism_evidence = validated_data.get("mechanism_evidence", None) # evidence is optional
+
+        try:
+            cv_mechanism_obj = CVMolecularMechanism.objects.get(
+                value = molecular_mechanism_value,
+                type = "mechanism"
+            )
+        except CVMolecularMechanism.DoesNotExist:
+            raise serializers.ValidationError({"message": f"Invalid mechanism value '{molecular_mechanism_value}'"})
+
+        try:
+            cv_support_obj = CVMolecularMechanism.objects.get(
+                value = molecular_mechanism_support,
+                type = "support"
+            )
+        except CVMolecularMechanism.DoesNotExist:
+            raise serializers.ValidationError({"message": f"Invalid mechanism support '{molecular_mechanism_support}'"})
+
+        # The mechanism synopsis is optional
+        cv_synopsis_obj = None
+        cv_synopsis_support_obj = None
+        if mechanism_synopsis is not None and "name" in mechanism_synopsis and mechanism_synopsis["name"] != "":
+            mechanism_synopsis_value = mechanism_synopsis.get("name")
+            mechanism_synopsis_support = mechanism_synopsis.get("support")
+
+            try:
+                cv_synopsis_obj = CVMolecularMechanism.objects.get(
+                    value = mechanism_synopsis_value,
+                    type = "mechanism_synopsis"
+                )
+            except CVMolecularMechanism.DoesNotExist:
+                raise serializers.ValidationError({"message": f"Invalid mechanism synopsis value '{mechanism_synopsis_value}'"})
+
+            try:
+                cv_synopsis_support_obj = CVMolecularMechanism.objects.get(
+                    value = mechanism_synopsis_support,
+                    type = "support"
+                )
+            except CVMolecularMechanism.DoesNotExist:
+                raise serializers.ValidationError({"message": f"Invalid mechanism synopsis support '{mechanism_synopsis_support}'"})
+
+        # Insert mechanism
+        mechanism_obj = MolecularMechanism.objects.create(
+            mechanism = cv_mechanism_obj,
+            mechanism_support = cv_support_obj,
+            synopsis = cv_synopsis_obj,
+            synopsis_support = cv_synopsis_support_obj,
+            is_deleted = 0
+        )
+
+        # Get evidence - the mechanism evidence was validated in the view 'LGDUpdateMechanism'
+        # Example: {'pmid': '25099252', 'description': 'text', 'evidence_types': 
+        #          [{'primary_type': 'Rescue', 'secondary_type': ['Human', 'Patient Cells']}]}
+        for evidence in mechanism_evidence:
+            pmid = evidence.get("pmid")
+
+            try:
```

Comment:
i think we should add the publication if it does not exist

Reply:
Yes, I added a note to do that in the future.